src/vsr/replica.zig

```diff
@@ -9516,7 +9478,9 @@ pub fn ReplicaType(
             assert(self.state_machine.pulse_needed(self.state_machine.prepare_timestamp));
 
             self.send_request_to_self(.pulse, &.{});
-            assert(self.pipeline.queue.contains_operation(.pulse));
+            if (!self.pipeline.queue.contains_operation(.pulse)) {
+                assert(self.view_durable_updating());
+            }
```

Comment:
As a rule of thumb, ff something might or might not happen, it is genreally better to assert that it does, and to push the condition where it doesn't up to the caller:

https://matklad.github.io/2023/11/15/push-ifs-up-and-fors-down.html

In this case, we actually already do that for upgrade requests!

Reply:
Ah, this a _great_ rule of thumb! It actually helped us  `assert(!self.view_durable_updating());` in primary_pipeline_prepare, send_request_pulse_to_self, _and_ send_request_upgrade_to_self. Added in https://github.com/tigerbeetle/tigerbeetle/pull/2253/commits/f0055581452f8c0ad86d7f2503316e650d863279.